File: 2023/9/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((line) => line.split(" ").map(Number));

function extrapolation(seq) {
  const arr = [seq];
  while (arr[arr.length - 1].some((a) => a != 0)) {
    const curr = arr[arr.length - 1];
    const next = [];
    for (let i = 1; i < curr.length; i++) {
      next.push(curr[i] - curr[i - 1]);
    }
    arr.push(next);
  }
  for (let i = arr.length - 2; i >= 0; i--) {
    arr[i].push(arr[i][arr[i].length - 1] + arr[i + 1][arr[i + 1].length - 1]);
    arr[i].unshift(arr[i][0] - arr[i + 1][0]);
  }
  return [arr[0][0], arr[0][arr[0].length - 1]];
}

const res = input.map(extrapolation);
const [part2, part1] = res.reduce((a, b) => [a[0] + b[0], a[1] + b[1]], [0, 0]);
console.log(part1, part2);

File: 2023/9/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data []string) [][]int {
	puzzleInput := make([][]int, len(data))
	for i, line := range data {
		s := strings.Split(line, " ")
		nums := make([]int, len(s))
		for j, n := range s {
			if num, err := strconv.Atoi(n); err != nil {
				panic(err)
			} else {
				nums[j] = num
			}
		}
		puzzleInput[i] = nums
	}
	return puzzleInput
}

func extrapolate(nums []int, nextValue bool) int {
	pyramid := [][]int{nums}
	for {
		curr := pyramid[len(pyramid)-1]
		next := make([]int, len(curr)-1)
		allZeros := true
		for i := range len(curr) - 1 {
			next[i] = curr[i+1] - curr[i]
			if next[i] != 0 {
				allZeros = false
			}
		}
		if allZeros {
			break
		}
		pyramid = append(pyramid, next)
	}
	res := 0
	for i := len(pyramid) - 1; i >= 0; i-- {
		if nextValue {
			res += pyramid[i][len(pyramid[i])-1]
		} else {
			res = pyramid[i][0] - res
		}
	}
	return res
}

func part1(puzzleInput [][]int) int {
	res := 0
	for _, nums := range puzzleInput {
		res += extrapolate(nums, true)
	}
	return res
}

func part2(puzzleInput [][]int) int {
	res := 0
	for _, nums := range puzzleInput {
		res += extrapolate(nums, false)
	}
	return res
}
